### PR TITLE
cmd_type: refactoring code to print type structs/enums/etc.

### PR DESCRIPTION
**Pull request by ret2libc**
This would benefit from https://github.com/radare/sdb/pull/166, so I can remove the code in cmd_type.c

r2r PR: https://github.com/radare/radare2-regressions/pull/1517

### CODE REVIEW

File: libr/core/cmd_type.c [hunk 1]
```diff
@@ -259,26 +259,57 @@ static void save_parsed_type(RCore *core, const char *parsed) {
...
+	SdbForeachCallback filter;
+	SdbList *list;
+};
+
+static int foreach_list_filter_cb(void *user, const char *k, const char *v) {
+	struct foreach_list_filter_t *u = (struct foreach_list_filter_t *)user;
+	SdbList *list = u->list;
+	if (!u->filter || u->filter (NULL, k, v)) {
+		SdbKv *kv = R_NEW0 (SdbKv);
+		kv->key = strdup (k);
```
Comment (radare) [inline libr/core/cmd_type.c]: check for null

File: libr/core/cmd_type.c [hunk 2]
```diff
@@ -259,26 +259,57 @@ static void save_parsed_type(RCore *core, const char *parsed) {
...
-//look at the next couple of functions
-//can be optimized into one right ... you see it you do it :P
-static int stdprintifstruct(void *p, const char *k, const char *v) {
-	if (!strncmp (v, "struct", strlen ("struct") + 1)) {
-		r_cons_println (k);
+// TODO these are going to be moved in sdb
+static int __cmp_asc(const void *a, const void *b) {
+	const SdbKv *ka = a;
+	const SdbKv *kb = b;
+	return strcmp (sdbkv_key (ka), sdbkv_key (kb));
```
Comment (radare) [inline libr/core/cmd_type.c]: you can cast in here, or have `*ka = a, *kb = b;` which is shorter

File: libr/core/cmd_type.c [hunk 3]
```diff
@@ -259,26 +259,57 @@ static void save_parsed_type(RCore *core, const char *parsed) {
...
+	SdbList *list;
+};
+
+static int foreach_list_filter_cb(void *user, const char *k, const char *v) {
+	struct foreach_list_filter_t *u = (struct foreach_list_filter_t *)user;
+	SdbList *list = u->list;
+	if (!u->filter || u->filter (NULL, k, v)) {
+		SdbKv *kv = R_NEW0 (SdbKv);
+		kv->key = strdup (k);
+		kv->value = strdup (v);
```
Comment (radare) [inline libr/core/cmd_type.c]: chk null?

File: libr/core/cmd_type.c [hunk 4]
```diff
@@ -386,40 +395,43 @@ static int sdbforcb_json (void *p, const char *k, const char *v) {
...
+
+	if (json) {
+		r_cons_printf ("[");
+	}
+	ls_foreach (l, it, kv) {
+		const char *k = sdbkv_key (kv);
+		if (!k || !*k) {
+			continue;
+		}
+
```
Comment (radare) [inline libr/core/cmd_type.c]: empty line

File: libr/core/cmd_type.c [hunk 5]
```diff
@@ -1211,26 +1220,22 @@ static int cmd_type(void *data, const char *input) {
...
-		case ' ':
-			{
-				char *res = sdb_querys (TDB, NULL, -1, sdb_fmt ("~~func.%s", input + 2));
-				r_cons_printf ("%s", res);
-				free (res);
-			//	r_core_cmdf (core, "tk~func.%s.", input + 2);
-			}
+		case ' ': {
+			char *res = sdb_querys (TDB, NULL, -1, sdb_fmt ("~~func.%s", input + 2));
+			r_cons_printf ("%s", res);
```
Comment (radare) [inline libr/core/cmd_type.c]: check for null

File: libr/core/cmd_type.c (hunk 1 above)
Comment (ret2libc) [reply libr/core/cmd_type.c]: I'd prefer if you can review these functions in the related sdb PR.

**Comment by ret2libc** [comment]
> can you add the C output too?

which C output?

**Comment by radare** [comment]
https://github.com/radare/radare2/issues/11756 <https://github.com/radare/radare2/issues/11756>

**Comment by radare** [comment]
yep

**Comment by ret2libc** [comment]
Mmm that would be better in a new PR. This small PR was just a refactoring to introduce sorting in all `t` commands that was necessary to make tests stable.

**Pull request merged**